### PR TITLE
[RHELC-1226] Host metering

### PR DESCRIPTION
**Pull request by r0x0d**
Jira Issues: [RHELC-1226](https://issues.redhat.com/browse/RHELC-1226)

Checklist

- [x] PR has been tested manually in a VM (either author or reviewer)
- [x] Jira issue has been made public if possible
- [x] `[RHELC-]` is part of the PR title 
- [x] GitHub label has been added to help with Release notes 
- [x] PR title explains the change from the user's point of view
- [x] Code and tests are documented properly
- [x] The commits are squashed to as few commits as possible (without losing data) 
- [ ] When merged: Jira issue has been updated to `Release Pending` if relevant

### CODE REVIEW

**Comment by r0x0d** [comment]
Picking up the work from Petr Vobornik. This is draft mode as I'm reading all the material to figure out what needs to be done to complete the PR.

File: convert2rhel/hostmetering.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+    if not conditions_met:
+        logger.info("Skipping host-metering configuration.")
+        return False
+
+    logger.info("Installing host-metering rpms.")
+    output, ret_install = call_yum_cmd("install", ["host-metering"])
+    logger.debug("Output of yum call: %s" % output)
+    if ret_install:
+        logger.warning("Failed to install host-metering rpms.")
```
Comment (SpyTec) [inline convert2rhel/hostmetering.py:88]: RPMs vs packages. I'd say we say packages as we use that elsewhere

File: convert2rhel/hostmetering.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+def _enable_host_metering_service():
+    logger.info("Enabling host-metering service.")
+    output, ret_enable = run_subprocess(["systemctl", "enable", "host-metering.service"])
+    logger.debug("Output of systemctl call: %s" % output)
+    if ret_enable:
+        logger.warning("Failed to enable host-metering service.")
+
+    logger.info("Starting host-metering service.")
+    output, ret_start = run_subprocess(["systemctl", "start", "host-metering.service"])
```
[the remark this answers is not in the document]
Comment (pr-watson) [reply convert2rhel/hostmetering.py:118]: I think returning False there would be good

File: convert2rhel/hostmetering.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    is_aws = rhsm_facts.get("aws_instance_id")
+    is_azure = rhsm_facts.get("azure_instance_id")
+    is_gcp = rhsm_facts.get("gcp_instance_id")
+    return any([is_aws, is_azure, is_gcp])
+
+
+def configure_host_metering():
+    """
+    Install, enable and start host-metering on the system when it is running
+    on a hyperscaller and is RHEL 7.
```
[the remark this answers is not in the document]
Comment (pr-watson) [reply convert2rhel/hostmetering.py]: Changed the doc string as the env var decides those steps and is not related to only RHEL 7 now

File: convert2rhel/hostmetering.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    Behavior can be controlled CONVERT2RHEL_CONFIGURE_HOST_METERING environment variable:
+    - "auto": host-metering will be configured based on the above conditions
+    - empty: host-metering will not be configured
+    - "force": forces configuration of host-metering (e.g., even if not running on a hyperscaller)
+    - any other value: behaves as empty
+
+    Returns:
+        bool: True if host-metering is configured successfully, False otherwise.
+    """
+    env_var = os.environ.get("CONVERT2RHEL_CONFIGURE_HOST_METERING")
```
Comment (r0x0d) [inline convert2rhel/hostmetering.py:69]: Let's default to the case where nothing happens 

```suggestion
    env_var = os.environ.get("CONVERT2RHEL_CONFIGURE_HOST_METERING", "empty")
```

File: convert2rhel/unit_tests/hostmetering_test.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__metaclass__ = type
+
+import pytest
+import six
+
+from convert2rhel import hostmetering, utils
```
Comment (r0x0d) [inline convert2rhel/unit_tests/hostmetering_test.py:23]:
```suggestion
from convert2rhel import hostmetering
```

File: convert2rhel/unit_tests/hostmetering_test.py [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        (
+            {"aws_instance_id": "i-1234567890abcdef0"},
+            Version(7, 9),
+            False,
+            "",  # option left empty
+        ),
+    ),
+)
+def test_hostmetering(monkeypatch, rhsm_facts, os_version, should_configure_metering, envvar):
+    if envvar is not None:
```
Comment (r0x0d) [inline convert2rhel/unit_tests/hostmetering_test.py:104]:
```suggestion
    if not envvar:
```

File: convert2rhel/unit_tests/hostmetering_test.py [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        monkeypatch.setenv("CONVERT2RHEL_CONFIGURE_HOST_METERING", envvar)
+
+    monkeypatch.setattr(system_info, "version", os_version)
+    monkeypatch.setattr(system_info, "releasever", "")  # reset as other test set it
```
Comment (r0x0d) [inline convert2rhel/unit_tests/hostmetering_test.py:108]: If other test set it globally, we should update the test to avoid test pollution down the road. 

We should fix as much test pollution as possible.

Comment (r0x0d) [reply convert2rhel/unit_tests/hostmetering_test.py:108]: Your unit_test should not reset the releasever because another test populated it wrongly, but rather, we should fix the test that pollutes it

File: convert2rhel/unit_tests/hostmetering_test.py [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        ),
+        (
+            {"aws_instance_id": "i-1234567890abcdef0"},
+            Version(7, 9),
+            False,
+            "",  # option left empty
+        ),
+    ),
+)
+def test_hostmetering(monkeypatch, rhsm_facts, os_version, should_configure_metering, envvar):
```
Comment (r0x0d) [inline convert2rhel/unit_tests/hostmetering_test.py:103]: Maybe we don't need to care about this right now given the deadline, but we could split this test into more tiny pieces to make it easier to work with.

File: tests/integration/tier1/destructive/host-metering/run_conversion.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+                "Enabling host-metering service.",
+                "Starting host-metering service.",
+            ]
+        )
+        assert c2r.expect("Conversion successful") == 0
+
+    assert c2r.exitstatus == 0
+
+    # There should not be any problems in coversion
+    assert shell("grep -i 'traceback' /var/log/convert2rhel/convert2rhel.log").returncode == 1
```
Comment (r0x0d) [inline tests/integration/tier1/destructive/host-metering/run_conversion.py:82]: @danmyway, maybe we don't need to check for this here? Not sure if we have a check for that at the end of all tests.

File: convert2rhel/unit_tests/hostmetering_test.py (hunk 7 above)
Comment (pr-watson) [reply convert2rhel/unit_tests/hostmetering_test.py:108]: I commented it out and the tests still pass. @pvoborni do you know what test is being referred to here?

File: convert2rhel/unit_tests/hostmetering_test.py (hunk 8 above)
Comment (pr-watson) [reply convert2rhel/unit_tests/hostmetering_test.py]: We can add that in the next release

File: convert2rhel/unit_tests/hostmetering_test.py (hunk 7 above)
Comment (pvoborni) [reply convert2rhel/unit_tests/hostmetering_test.py]: IIRC (was in December), I got a test failure without it. But unfortunately I don't remember more. So if it passes without then it can be removed.

File: convert2rhel/subscription.py [hunk 10]
```diff
@@ -64,6 +68,16 @@ class StopRhsmError(Exception):
...
 
 
+class FileNotFoundError(OSError):
+    """Raised when the requested file cannot be found."""
+
+    pass
+
+
+class JSONDecodeError(ValueError):
+    pass
```
Comment (r0x0d) [inline convert2rhel/subscription.py:78]: Why do we need to recreate exceptions that already exist? Is this because python 2 or something similar?

Comment (pr-watson) [reply convert2rhel/subscription.py]: Yes for centos7 the tests were failing for those exceptions, but it had something to do with me using tmp_path instead of tmp_dir. Tests are passing now so i'm removing them

**Comment by pr-watson** [comment]
> We still have some occurrences of the word `hyperscaller`:
> 
> ```
> convert2rhel/hostmetering.py
> 40:    Check if the system is running on hyperscaller. Currently supported
> 41:    hyperscallers are aws, azure and gcp.
> 45:    :return: True if the system is running on hyperscaller, False otherwise.
> 62:    - "force": forces configuration of host-metering (e.g., even if not running on a hyperscaller)
> 
> tests/integration/tier1/destructive/host-metering/test_host_metering.py
> 25:    hyperscaller.
> 
> tests/integration/tier1/destructive/host-metering/run_conversion.py
> 58:    service on hyperscallers on RHEL 7.9.
> 
> convert2rhel/unit_tests/hostmetering_test.py
> 38:            False,  # not on hyperscaller
> ```

All instances have now been updated

File: convert2rhel/unit_tests/hostmetering_test.py (hunk 7 above)
Comment (r0x0d) [reply convert2rhel/unit_tests/hostmetering_test.py]: Possibly it was another test pollution we had, but maybe got solved with the recent merges then. I fixed a couple of them in some of my PRs that got merged.

File: convert2rhel/subscription.py [hunk 11]
```diff
@@ -902,6 +906,29 @@ def update_rhsm_custom_facts():
...
+    try:
+        with open(RHSM_FACTS_FILE, mode="r") as handler:
+            rhsm_facts = json.load(handler)
+            loggerinst.info("RHSM facts collected successfully.")
+            return rhsm_facts
+    except (FileNotFoundError, json.JSONDecodeError) as e:
+        loggerinst.critical_no_exit(
+            "Failed to get the RHSM facts with return code '%s' and output '%s'.",
+        )
+        return rhsm_facts
```
[the remark this answers is not in the document]
Comment (pr-watson) [reply convert2rhel/subscription.py]: Added tests

File: convert2rhel/hostmetering.py [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+    - any other value: behaves as empty
+
+    :return: True if host-metering is configured successfully, False otherwise.
+    :rtype: bool
+    """
+    env_var = os.environ.get("CONVERT2RHEL_CONFIGURE_HOST_METERING", "empty")
+    if "CONVERT2RHEL_CONFIGURE_HOST_METERING" not in os.environ:
+        return False
+
+    if system_info.version.major != 7 and env_var != "force":
```
Comment (r0x0d) [inline convert2rhel/hostmetering.py:72]: Do we need to check for env_var != "force" here?

File: convert2rhel/subscription.py [hunk 13]
```diff
@@ -902,6 +906,30 @@ def update_rhsm_custom_facts():
...
+    try:
+        with open(RHSM_FACTS_FILE, mode="r") as handler:
+            rhsm_facts = json.load(handler)
+            loggerinst.info("RHSM facts loaded.")
+            return rhsm_facts
+    except (FileNotFoundError, json.JSONDecodeError) as e:
+        loggerinst.critical_no_exit(
+            "Failed to get the RHSM facts : %s." % e,
+        )
+        return rhsm_facts
```
Comment (r0x0d) [inline convert2rhel/subscription.py:930]: Not a blocker, but we could do: 
```suggestion
    loggerinst.info("Reading RHSM facts file.")
    try:
        with open(RHSM_FACTS_FILE, mode="r") as handler:
            rhsm_facts = json.load(handler)
            loggerinst.info("RHSM facts loaded.")
            return rhsm_facts
    except (FileNotFoundError, json.JSONDecodeError) as e:
        loggerinst.critical_no_exit(
            "Failed to get the RHSM facts : %s." % e,
        )
     finally:
        return rhsm_facts
```

File: convert2rhel/hostmetering.py (hunk 12 above)
Comment (pr-watson) [reply convert2rhel/hostmetering.py:72]: Since the force option will configure host metering unconditionally I feel it's required

File: plans/tier1.fmf [hunk 14]
```diff
@@ -249,3 +249,20 @@ description+: |
         discover+:
             test+<:
                 - conversion-method/rhsm_conversion
+
+    /host-metering:
+        enabled: false
+        adjust+:
+            # TODO: enable only when running on AWS, GCP, Azure
```
Comment (r0x0d) [inline plans/tier1.fmf:256]: @danmyway @kokesak is there anything we need to do here?

File: tests/integration/tier1/destructive/host-metering/main.fmf [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+    Verify that:
+    - host-metering service is enabled and started
+
+link:
+    - https://issues.redhat.com/browse/RHELC-1226
+
+tag+:
+    - host-metering
+
+test: pytest -svv -m test_host_metering
```
Comment (r0x0d) [inline tests/integration/tier1/destructive/host-metering/main.fmf:15]: Let's update the content here to get rid of --payg mention and use the env var

File: convert2rhel/hostmetering.py (hunk 12 above)
Comment (r0x0d) [reply convert2rhel/hostmetering.py:72]: sounds good

**Review by r0x0d — COMMENTED** [review]
It looks good to me. I think we have to test this somewhere, but looks very good.

File: convert2rhel/subscription.py [hunk 16]
```diff
@@ -902,6 +906,30 @@ def update_rhsm_custom_facts():
...
+    try:
+        with open(RHSM_FACTS_FILE, mode="r") as handler:
+            rhsm_facts = json.load(handler)
+            loggerinst.info("RHSM facts loaded.")
+    except (FileNotFoundError, json.JSONDecodeError) as e:
+        loggerinst.critical_no_exit(
+            "Failed to get the RHSM facts : %s." % e,
+        )
+     finally:
+        return rhsm_facts
```
Comment (SpyTec) [inline convert2rhel/subscription.py:930]: There is an extra space on finally here, but also we don't need `finally` AFAIK
```suggestion
    loggerinst.info("Reading RHSM facts file.")
    try:
        with open(RHSM_FACTS_FILE, mode="r") as handler:
            rhsm_facts = json.load(handler)
            loggerinst.info("RHSM facts loaded.")
    except (FileNotFoundError, json.JSONDecodeError) as e:
        loggerinst.critical_no_exit(
            "Failed to get the RHSM facts : %s." % e,
        )
    return rhsm_facts
```

File: convert2rhel/hostmetering.py [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+            " For generic usage please use the 'auto' option."
+        )
+    elif env_var == "auto":
+        should_configure_metering = True
+    else:
+        should_configure_metering = False
+
+    rhsm_facts = get_rhsm_facts()
+    is_hyperscaler = is_running_on_hyperscaler(rhsm_facts)
+    if (is_hyperscaler not  or not should_configure_metering) and env_var != "force":
```
[the remark this answers is not in the document]
Comment (SpyTec) [reply convert2rhel/hostmetering.py:90]: Yes, this is an issue
```suggestion
    if (is_hyperscaler or not should_configure_metering) and env_var != "force":
```

**Review by SpyTec — CHANGES_REQUESTED** [review]

File: plans/tier1.fmf (hunk 14 above)
Comment (pvoborni) [reply plans/tier1.fmf:256]: This might not be needed as the test is setting `    os.environ["CONVERT2RHEL_CONFIGURE_HOST_METERING"] = "force"` and using fake endpoint (localhost).

File: convert2rhel/subscription.py [hunk 18]
```diff
@@ -902,6 +906,30 @@ def update_rhsm_custom_facts():
...
+    loggerinst.info("Reading RHSM facts file.")
+    try:
+        with open(RHSM_FACTS_FILE, mode="r") as handler:
+            rhsm_facts = json.load(handler)
+            loggerinst.info("RHSM facts loaded.")
+    except (FileNotFoundError, json.JSONDecodeError) as e:
+        loggerinst.critical_no_exit(
+            "Failed to get the RHSM facts : %s." % e,
+        )
+     finally:
```
[the remark this answers is not in the document]
Comment (hosekadam) [reply convert2rhel/subscription.py:929]: I see Freya mentioned that, so this is duplicate

**Comment by pr-watson** [comment]
I'll remove the finally

**Comment by pr-watson** [comment]
Had to update FileNotFoundError and JsonDecodeError to IOError and ValueError respectively because python 2 is not recognizing them. JsonDecodeError just raises a ValueError so it's virtually the same

**Review by SpyTec — APPROVED** [review]
Changes I requested were addressed

File: convert2rhel/unit_tests/subscription_test.py [hunk 19]
```diff
@@ -1266,3 +1267,62 @@ def test_update_rhsm_custom_facts_disable_telemetry(monkeypatch, caplog):
...
+        caplog.records[-1].message
+        == "Failed to get the RHSM facts : [Errno 2] No such file or directory: 'fake/location/file.json'."
+    )
+    assert isinstance(facts, dict)
+    assert len(facts) == 0
+    assert run_mock.call_count == 0
+
+
+@centos7
+def test_get_rhsm_facts_json_decode_error(monkeypatch, global_tool_opts, caplog, tmpdir, pretend_os):
```
Comment (SpyTec) [inline convert2rhel/unit_tests/subscription_test.py:1316]: Could we not encounter this on anything other than centos7? The other tests don't test specifically for centos7

Comment (pr-watson) [reply convert2rhel/unit_tests/subscription_test.py]: I added that because el7 systems will only hit that specific message, for el8 the message differs slightly so I made another test for that

**Comment by SpyTec** [comment]
Integration tests seems to be related to change

```
Traceback (most recent call last):


  File "/usr/bin/convert2rhel", line 11, in <module>


    load_entry_point('convert2rhel==1.6.1', 'console_scripts', 'convert2rhel')()


  File "/usr/lib/python3.6/site-packages/convert2rhel/initialize.py", line 80, in run


    from convert2rhel import main


  File "/usr/lib/python3.6/site-packages/convert2rhel/main.py", line 23, in <module>


    from convert2rhel import actions, applock, backup, breadcrumbs, checks, exceptions, grub, hostmetering


  File "/usr/lib/python3.6/site-packages/convert2rhel/hostmetering.py", line 30, in <module>


    from convert2rhel.subscription import get_rhsm_facts


  File "/usr/lib/python3.6/site-packages/convert2rhel/subscription.py", line 34, in <module>


    from convert2rhel.redhatrelease import os_release_file


  File "/usr/lib/python3.6/site-packages/convert2rhel/redhatrelease.py", line 95, in <module>


    system_release_file = RestorableFile(get_system_release_filepath())  # pylint: disable=C0103


  File "/usr/lib/python3.6/site-packages/convert2rhel/backup/files.py", line 41, in __init__


    if not os.path.isabs(filepath):


  File "/usr/lib64/python3.6/posixpath.py", line 66, in isabs


    s = os.fspath(s)


TypeError: expected str, bytes or os.PathLike object, not NoneType



```

File: convert2rhel/unit_tests/subscription_test.py [hunk 20]
```diff
@@ -1266,3 +1267,80 @@ def test_update_rhsm_custom_facts_disable_telemetry(monkeypatch, caplog):
...
+    assert caplog.records[-1].levelname == "CRITICAL"
+    assert caplog.records[-1].message == "Failed to get the RHSM facts : No JSON object could be decoded."
+    assert isinstance(facts, dict)
+    assert len(facts) == 0
+    assert run_mock.call_count == 0
+
+
+@centos8
+@pytest.mark.skipif(pkgmanager.TYPE == "yum", reason="Test is only relevant for RHEL 8+")
+def test_get_rhsm_facts_json_decode_error_el7(monkeypatch, global_tool_opts, caplog, tmpdir, pretend_os):
```
Comment (r0x0d) [inline convert2rhel/unit_tests/subscription_test.py:1334]: Could we update the name in here to el8?

```suggestion
def test_get_rhsm_facts_json_decode_error_el8(monkeypatch, global_tool_opts, caplog, tmpdir, pretend_os):
```

File: plans/tier1.fmf (hunk 14 above)
Comment (kokesak) [reply plans/tier1.fmf:256]: All the images used are in AWS. Only that centos-7 and centos-8 images are our own there (not publicly available in the AWS marketplace) the rest are the official ones. I think we don't need this condition at all.

File: convert2rhel/hostmetering.py [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+
+
+def configure_host_metering():
+    """
+    Decide whether to install, enable and start host-metering on the system based on the
+    CONVERT2RHEL_CONFIGURE_HOST_METERING environment variable.
+
+    Behavior can be controlled CONVERT2RHEL_CONFIGURE_HOST_METERING environment variable:
+    - "auto": host-metering will be configured based on the above conditions
+    - empty: host-metering will not be configured
```
Comment (r0x0d) [inline convert2rhel/hostmetering.py:61]: Probably we could remove this from here, or reword it to say that if the env var is not present, nothing will happen. Having the keyword `empty` here caused me a confusion to think that we could potentially say `CONVERT2RHEL_CONFIGURE_HOST_METERING=empty`.

File: convert2rhel/hostmetering.py [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+    - "force": forces configuration of host-metering (e.g., even if not running on a hyperscaler)
+    - any other value: behaves as empty
```
Comment (r0x0d) [inline convert2rhel/hostmetering.py:63]:
```suggestion
    - any other value: Will be ignored and host metering will not be configured.
```

File: convert2rhel/hostmetering.py [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+
+    :return: True if host-metering is configured successfully, False otherwise.
+    :rtype: bool
+    """
+    env_var = os.environ.get("CONVERT2RHEL_CONFIGURE_HOST_METERING", "empty")
+    if "CONVERT2RHEL_CONFIGURE_HOST_METERING" not in os.environ:
+        return False
```
Comment (r0x0d) [inline convert2rhel/hostmetering.py:70]: Following the the idea of the docstring above that mentions `any other value`, maybe we should improve this condition? Probably something like: 

```suggestion
    if env_var not in ("force", "auto"):
        return False
```

File: plans/tier1.fmf [hunk 24]
```diff
@@ -249,3 +249,20 @@ description+: |
...
+        discover+:
+            filter:
+            - tag:checks-after-conversion | tag:host-metering
+        prepare+:
+            - name: test host-metering
+              how: shell
+              script: pytest -svv tests/integration/*/destructive/host-metering/run_conversion.py
+            - name: reboot after conversion
+              how: ansible
+              playbook: tests/ansible_collections/roles/reboot/main.yml
```
Comment (kokesak) [inline plans/tier1.fmf:268]: We actually do not use this anymore. The whole `prepare` should not be used anymore. It should look like this:
```
      discover+:
            test+<:
                - host-metering/run-host-metering-conversion
            test+:
                - host-metering/test-active-host-metering
```
Then the "tests/integration/tier1/destructive/host-metering/main.fmf" should contain this
```
/run-host-metering-conversion:
    tag+:
        - test-host-metering-conversion
    test: pytest -svv -m test_host_metering

/test-active-host-metering:
    tag+:
       - test-active-host-metering
    test: pytest -svv -m test_active_host_metering
```

Comment (pr-watson) [reply plans/tier1.fmf:268]: @kokesak yes that's okay

File: tests/integration/tier1/destructive/host-metering/run_conversion.py (hunk 9 above)
Comment (danmyway) [reply tests/integration/tier1/destructive/host-metering/run_conversion.py:82]: I think you have rather meant [this test](https://github.com/oamg/convert2rhel/blob/main/tests/integration/common/checks-after-conversion/test_verify_strings_in_log.py) @kokesak and yes, it would make sense to have it as an after conversion check.
However, I just wonder, @r0x0d, is it a valid scenario to have a traceback in the log and the conversion finishing successfully at the same time?

Comment (r0x0d) [reply tests/integration/tier1/destructive/host-metering/run_conversion.py:82]: It is more likely to have a traceback during an analysis rather than a conversion.

File: plans/tier1.fmf (hunk 24 above)
Comment (kokesak) [reply plans/tier1.fmf:268]: Thanks @danmyway for trying that

File: tests/integration/tier1/destructive/host-metering/run_conversion.py [hunk 25]
```diff
@@ -0,0 +1,82 @@
...
+        c2r.expect(
+            [
+                "Installing host-metering rpms.",
+                "Enabling host-metering service.",
+                "Starting host-metering service.",
+            ]
+        )
+        assert c2r.expect("Conversion successful") == 0
+
+    assert c2r.exitstatus == 0
```
[the remark this answers is not in the document]
Comment (r0x0d) [reply tests/integration/tier1/destructive/host-metering/run_conversion.py]: Yeah, I think that's rigth

**Review by bookwar — APPROVED** [review]
